### ISSUE TITLE
Layout of single-variant enums

### ISSUE THREAD

**Issue by nikomatsakis** [rust-rfcs/unsafe-code-guidelines#79]
Enums that contain a **single variant** and which do not have an
explicit `#[repr]` annotation are an important special case. Since
there is only a single variant, the enum must be instantiated with
that variant, which means that the enum is in fact equivalent to a
struct. The question then is to what extent we should **guarantee**
that the two share an equivalent layout, and also how to define the
interaction with uninhabited types.

As presently implemented, the compiler will use the same layout for
structs and for single variant enums (as long as they do not have a
`#[repr]` annotation that overrides that choice). So, for example, the
struct `SomeStruct` and the enum `SomeEnum` would have an equivalent
layout ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3697ac684c3d021892694956df957653))::

```rust
struct SomeStruct;
enum SomeEnum {
  SomeVariant,
}
```

Similarly, the struct `SomeStruct` and the enum `SomeVariant` in this
example would also be equivalent in their layout
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=924724764419f846c788a8763da45992)):

```rust
struct SomeStruct { x: u32 }
enum SomeEnum {
  SomeVariant { x: u32 },
}
```

In fact, the compiler will use this optimized layout even for enums
that define multiple variants, as long as all but one of the variants
is uninhabited
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3cc1484c5b91097f3dc2015b7c207a0e)):

```rust
struct SomeStruct { x: u32 }
enum SomeEnum {
  SomeVariant { x: u32 },
  UninhabitedVariant { y: Void },
}
```

**Comment by gnzlbg** [comment]
> In fact, the compiler will use this optimized layout even for enums
that define multiple variants, as long as all but one of the variants
is uninhabited

We could resolve this with:

---

The layout of **multi-variant** enums with **one inhabited variant** is the same
as that of the single-variant enum containing that same inhabited variant.

For example, here:

```rust
struct SomeStruct { x: u32 }
enum SingleVariantDataCarrying {
    DataCarryingVariant(SomeStruct),
}
enum Void0 {}
enum Void1 {}
enum MultiVariantSingleInhabited {
  DataCarryingVariant(SomeStruct),
  Uninhabited0(Void0),
  Uninhabited1(Void1),
}
```

the layout of `MultiVariantSingleInhabited` is the same layout as that of
`SingleVariantDataCarrying`.

---

but doing so is blocked on resolving the layout of uninhabited types: #165 .

**Comment by rkruppe** [comment]
That's not what is currently (or is planned to be) [implemented](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=72f4311b1d71d65d074dd76ff0cca6ac), again because uninhabited types aren't all ZSTs and can be partially constructed. The [plan of record](https://github.com/rust-lang/rust/issues/49298#issuecomment-380844923) is to only omit variants which are uninhabited *and ZSTs*.

**Comment by gnzlbg** [comment]
@rkruppe Makes sense. I think this might be then covered with a guarantee about layout for unions with a single (inhabited?) non-zero-sized field.

**Comment by RalfJung** [comment]
@rkruppe those issues apply to structs though, not to enums. Surface Rust does not allow partial initialization of enums. But maybe MIR does... in the spirit of https://github.com/rust-lang/rust/issues/63616, I wonder what happens with something like
```
let x: Result<u32, (u32, !)> = Err((42, panic!()));
```

**Comment by rkruppe** [comment]
Yes, that will have partial initialization in MIR. "Surface Rust allows partial initialization" was always just part of the story (and is even ruled out again right now!), the MIR-level motivation, specifically enabling aggressive move forwarding/emplacement without having to special case siblings of uninhabited places (to introduce fake temporary destinations), applies universally to structs, enums, and every other aggregate.

**Comment by heaths** [comment]
Seems there was some [discussion](https://github.com/rust-lang/rust/issues/60405#issuecomment-572812900) in 2020 about univariant enums. What I cant find is if there was ever a discussion - or, better, a decision - on supporting `#[repr(transparent)]` on bivariant enums with an empty variant like `Result<(), u32>` or similar to cover [FFI-compatible scenarios](https://github.com/rust-lang/rust/issues/55615). @gnzlbg I see your moniker related to a lot of this discussion. Do you know? From the aforementioned issue, I didn't see any follow-up since 2019 about an RFC. Is that still needed?

**Comment by scottmcm** [comment]
@heaths It's unclear to me what you expect transparent on a multi-variant enum like `Result<(), u32>` is supposed to mean -- that type is 64 bits as it needs a discriminant and the payload.  Did you mean something where one of the arms is impossible, like `Result<u32, !>`?

**Comment by heaths** [comment]
I see your point. Perhaps this is overly optimizing for a specific use case - FFI - but since my (and others') intention was to define native functions returning non-zero errors, would constraints make it easier to discriminate e.g. `Result<(), std::num::NonZeroU32>`?

**Comment by scottmcm** [comment]
@heaths That's a niche optimization question, so you're probably looking for a different issue -- it's definitely not a "single-variant enum", even if it might layout-optimize down to 32 bits.

**Comment by heaths** [comment]
I ran across https://github.com/rust-lang/rust/issues/84277 as well, which eventually - hopefully - is a way to solve this seemingly common FFI scenario. I agree it's a niche optimization, though.

**Comment by scottmcm** [comment]
@heaths Yes, that exact FFI scenario is written out specifically in the RFC: https://rust-lang.github.io/rfcs/3058-try-trait-v2.html#implementing-try-for-a-non-generic-type

**Comment by digama0** [comment]
By the way, from this exchange I think there might have been a miscommunication: @scottmcm means it's a (niche optimization) question, not a niche (optimization question). That is, it's a question about optimization of niches, not a question about optimizations which are applicable only to a restricted set of scenarios.

**Comment by pnkfelix** [comment]
Discussed at T-opsem backlog bonanza.

This is T-lang, not opsem

@rustbot label: +T-lang

**Comment by zachs18** [comment]
(linking related discussion) For the special case of a single-variant `repr(Rust)` enum with no fields or with only zero-sized fields, https://github.com/rust-lang/reference/pull/2262 will guarantee that it is zero-sized (once T-lang FCP completes).
